# Cleaning plan for user data

In [18]:
import pandas as pd

### Nulls?

In [19]:
users = pd.read_csv("../data/raw/users_data.csv")

users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5300 entries, 0 to 5299
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   user_id            5150 non-null   object 
 1   first_name         5150 non-null   object 
 2   last_name          5150 non-null   object 
 3   age                5150 non-null   float64
 4   gender             5150 non-null   object 
 5   birthdate          5150 non-null   object 
 6   sign_up_date       5150 non-null   object 
 7   user_location      5150 non-null   object 
 8   subscription_plan  5150 non-null   object 
dtypes: float64(1), object(8)
memory usage: 372.8+ KB


In [20]:
# Check for nulls
print("Num of nulls: " + str(users.isnull().any(axis=1).sum()))
print(users.shape)

Num of nulls: 150
(5300, 9)


In [21]:
print("Number of missing values per column")
for col in users.columns:
    print(f"{col}: {users[col].isnull().sum()}")

Number of missing values per column
user_id: 150
first_name: 150
last_name: 150
age: 150
gender: 150
birthdate: 150
sign_up_date: 150
user_location: 150
subscription_plan: 150


> Probably have entire null rows because the columns have the same amount of nulls...

In [22]:
# Finds rows with nulls
users = users.replace(" ", None)
users_nulls = users[users.isna().any(axis=1)]
print(users_nulls)

     user_id first_name last_name  age gender birthdate sign_up_date  \
7        NaN        NaN       NaN  NaN    NaN       NaN          NaN   
9        NaN        NaN       NaN  NaN    NaN       NaN          NaN   
53       NaN        NaN       NaN  NaN    NaN       NaN          NaN   
243      NaN        NaN       NaN  NaN    NaN       NaN          NaN   
340      NaN        NaN       NaN  NaN    NaN       NaN          NaN   
...      ...        ...       ...  ...    ...       ...          ...   
5239     NaN        NaN       NaN  NaN    NaN       NaN          NaN   
5243     NaN        NaN       NaN  NaN    NaN       NaN          NaN   
5254     NaN        NaN       NaN  NaN    NaN       NaN          NaN   
5283     NaN        NaN       NaN  NaN    NaN       NaN          NaN   
5291     NaN        NaN       NaN  NaN    NaN       NaN          NaN   

     user_location subscription_plan  
7              NaN               NaN  
9              NaN               NaN  
53             NaN

> 150 ~ 2.8% -> Not a lot of rows and only fully blank rows thus it can be removed

In [23]:
users = users.dropna()
print("Num of nulls: " + str(users.isnull().any(axis=1).sum()))
print(users.shape)

Num of nulls: 0
(5150, 9)


### Correct data types?

In [24]:
users = pd.read_csv("../data/raw/users_data.csv")

users = users.replace(" ", None)
users = users.dropna()
users.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5150 entries, 0 to 5299
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   user_id            5150 non-null   object 
 1   first_name         5150 non-null   object 
 2   last_name          5150 non-null   object 
 3   age                5150 non-null   float64
 4   gender             5150 non-null   object 
 5   birthdate          5150 non-null   object 
 6   sign_up_date       5150 non-null   object 
 7   user_location      5150 non-null   object 
 8   subscription_plan  5150 non-null   object 
dtypes: float64(1), object(8)
memory usage: 402.3+ KB


In [25]:
users.head()

user_id first_name last_name   age  gender   birthdate sign_up_date  \
0  user_1595       Mark  Phillips  55.0    Male  12/07/2001   20/03/2023   
1  user_3596      Karen     Baker  52.0  Female  02/07/2005   21/06/2023   
2    user_25    Brandon    Walker  50.0    Male  24/01/1987   09/09/2022   
3   user_403     Daniel    Howard  39.0    Male  16/02/1998   22/04/2022   
4  user_1213  Stephanie     Lewis  24.0  Female  29/08/1967   24/10/2022   

   user_location subscription_plan  
0        Seattle               Pro  
1         Denver             Basic  
2  San Francisco           Student  
3         Denver           Student  
4         Boston             Basic

In [26]:
users = users.astype(
    {"age": int, "birthdate": "datetime64[ns]", "sign_up_date": "datetime64[ns]"}
)

print(users.info())
users.head()

<class 'pandas.core.frame.DataFrame'>
Index: 5150 entries, 0 to 5299
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   user_id            5150 non-null   object        
 1   first_name         5150 non-null   object        
 2   last_name          5150 non-null   object        
 3   age                5150 non-null   int64         
 4   gender             5150 non-null   object        
 5   birthdate          5150 non-null   datetime64[ns]
 6   sign_up_date       5150 non-null   datetime64[ns]
 7   user_location      5150 non-null   object        
 8   subscription_plan  5150 non-null   object        
dtypes: datetime64[ns](2), int64(1), object(6)
memory usage: 402.3+ KB
None


user_id first_name last_name  age  gender  birthdate sign_up_date  \
0  user_1595       Mark  Phillips   55    Male 2001-12-07   2023-03-20   
1  user_3596      Karen     Baker   52  Female 2005-02-07   2023-06-21   
2    user_25    Brandon    Walker   50    Male 1987-01-24   2022-09-09   
3   user_403     Daniel    Howard   39    Male 1998-02-16   2022-04-22   
4  user_1213  Stephanie     Lewis   24  Female 1967-08-29   2022-10-24   

   user_location subscription_plan  
0        Seattle               Pro  
1         Denver             Basic  
2  San Francisco           Student  
3         Denver           Student  
4         Boston             Basic

### Duplicates?

In [27]:
users = pd.read_csv("../data/raw/users_data.csv")

users = users.replace(" ", None)
users = users.dropna()
users = users.astype(
    {
        "age": int,
        "birthdate": "datetime64[ns]",
        "sign_up_date": "datetime64[ns]",
    }
)
users.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5150 entries, 0 to 5299
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   user_id            5150 non-null   object        
 1   first_name         5150 non-null   object        
 2   last_name          5150 non-null   object        
 3   age                5150 non-null   int64         
 4   gender             5150 non-null   object        
 5   birthdate          5150 non-null   datetime64[ns]
 6   sign_up_date       5150 non-null   datetime64[ns]
 7   user_location      5150 non-null   object        
 8   subscription_plan  5150 non-null   object        
dtypes: datetime64[ns](2), int64(1), object(6)
memory usage: 402.3+ KB


In [28]:
print("Num of duplicates: " + str(users.duplicated().sum()))
print(users[users.duplicated()])

Num of duplicates: 150
        user_id first_name last_name  age  gender  birthdate sign_up_date  \
120   user_1325      Betty    Turner   31  Female 1977-08-09   2022-09-15   
397   user_1988    Cynthia    Harris   54  Female 1961-02-27   2021-12-11   
502   user_4061  Elizabeth     Young   38  Female 1989-12-29   2023-05-14   
536   user_2874     Sandra     Lewis   64  Female 2001-04-08   2022-06-14   
862   user_4172     Angela  Gonzalez   37  Female 1978-07-23   2021-03-26   
...         ...        ...       ...  ...     ...        ...          ...   
5206  user_4275  Katherine      Hill   64  Female 1999-09-16   2023-03-25   
5213  user_4467      Linda  Robinson   33  Female 1975-03-23   2023-11-09   
5229  user_1447     Andrew      Diaz   23    Male 2003-02-28   2023-04-23   
5260  user_1120     Thomas     Young   22    Male 1997-04-09   2022-11-22   
5289  user_1768     Justin     Davis   34    Male 1990-06-02   2021-08-23   

      user_location subscription_plan  
120         

In [29]:
users = users.drop_duplicates()
print("Num of duplicates: " + str(users.duplicated().sum()))
print(users.shape)

Num of duplicates: 0
(5000, 9)


### Looking at dates

In [30]:
users = pd.read_csv("../data/raw/users_data.csv")

users = users.replace(" ", None)
users = users.dropna()
users = users.astype(
    {
        "age": int,
        "birthdate": "datetime64[ns]",
        "sign_up_date": "datetime64[ns]",
    }
)
users = users.drop_duplicates()
users.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5000 entries, 0 to 5299
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   user_id            5000 non-null   object        
 1   first_name         5000 non-null   object        
 2   last_name          5000 non-null   object        
 3   age                5000 non-null   int64         
 4   gender             5000 non-null   object        
 5   birthdate          5000 non-null   datetime64[ns]
 6   sign_up_date       5000 non-null   datetime64[ns]
 7   user_location      5000 non-null   object        
 8   subscription_plan  5000 non-null   object        
dtypes: datetime64[ns](2), int64(1), object(6)
memory usage: 390.6+ KB


In [31]:
users.head()

user_id first_name last_name  age  gender  birthdate sign_up_date  \
0  user_1595       Mark  Phillips   55    Male 2001-12-07   2023-03-20   
1  user_3596      Karen     Baker   52  Female 2005-02-07   2023-06-21   
2    user_25    Brandon    Walker   50    Male 1987-01-24   2022-09-09   
3   user_403     Daniel    Howard   39    Male 1998-02-16   2022-04-22   
4  user_1213  Stephanie     Lewis   24  Female 1967-08-29   2022-10-24   

   user_location subscription_plan  
0        Seattle               Pro  
1         Denver             Basic  
2  San Francisco           Student  
3         Denver           Student  
4         Boston             Basic

> The birthdates and ages don't line up likely due to it not being specified during data generation. -> Going to assume the age is correct and discard their birthdates as I don't need them for my visualisations

In [32]:
users = users.drop("birthdate", axis=1)
users.head()
users.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5000 entries, 0 to 5299
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   user_id            5000 non-null   object        
 1   first_name         5000 non-null   object        
 2   last_name          5000 non-null   object        
 3   age                5000 non-null   int64         
 4   gender             5000 non-null   object        
 5   sign_up_date       5000 non-null   datetime64[ns]
 6   user_location      5000 non-null   object        
 7   subscription_plan  5000 non-null   object        
dtypes: datetime64[ns](1), int64(1), object(6)
memory usage: 351.6+ KB


### Save test data

In [35]:
users = pd.read_csv("../data/raw/users_data.csv")

users = users.replace(" ", None)
users = users.dropna()
users = users.astype(
    {
        "age": int,
        "birthdate": "datetime64[ns]",
        "sign_up_date": "datetime64[ns]",
    }
)
users = users.drop_duplicates()
users = users.drop("birthdate", axis=1)

users.to_csv("../tests/test_data/expected/users.csv", index=False)